aeppl/opt.py

```diff
@@ -178,8 +185,136 @@ def naive_bcast_rv_lift(fgraph, node):
     return [bcasted_node.outputs[1]]
 
 
+class MixtureRV(RandomVariable):
+    """A placeholder used to specify a log-likelihood for a mixture sub-graph."""
+
+    def _shape_from_params(self, dist_params, param_shapes=None, **kwargs):
+        if param_shapes is not None:
+            return at.broadcast_shape(*dist_params[1:])
+        else:
+            return at.broadcast_shape(*param_shapes, arrays_are_shapes=True)
+
+
+@local_optimizer(subtensor_ops)
+def mixture_replace(fgraph, node):
+    """Identify mixture sub-graphs and replace them with a place-holder `Op`."""
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return
+
+    if not isinstance(node.op, subtensor_ops):
+        return
+
+    out_var = node.outputs[0]
+
+    # These should be a combination of `RandomVariable`s
+    joined_rvs = node.inputs[0]
+
+    # First, make sure that it's some sort of combination
+    if not (joined_rvs.owner and isinstance(joined_rvs.owner.op, (MakeVector, Join))):
+        return
+
+    # TODO: Confirm that a log-likelihood can be computed for these arguments
+    if isinstance(joined_rvs.owner.op, MakeVector):
+        mixture_rvs = joined_rvs.owner.inputs
+
+    elif isinstance(joined_rvs.owner.op, Join):
+        # TODO: Will we need this in the `logprob` function?
+        # join_axis = joined_rvs.owner.inputs[0]
+        mixture_rvs = joined_rvs.owner.inputs[1:]
+
+    if not all(
+        rv.owner and isinstance(rv.owner.op, RandomVariable) for rv in mixture_rvs
+    ):
+        # TODO: Allow constants and make them Dirac-deltas
+        # Not supported
+        return
+
+    indices = indices_from_subtensor(
+        getattr(node.op, "idx_list", None), node.inputs[1:]
+    )
+
+    if len(indices) == 1:
+        if not indices[0].owner:
+            # TODO: This could be a `Constant`; we might want to support this
+            return
+
+        if isinstance(indices[0].owner.op, ScalarFromTensor):
+            # This covers the case of a single scalar index
+            mixing_rv = indices[0].owner.inputs[0]
+
+            if not (mixing_rv.owner and isinstance(mixing_rv.owner.op, RandomVariable)):
+                # Not supported
+                return
+
+        elif isinstance(indices[0].owner.op, RandomVariable):
+            mixing_rv = indices[0]
+        else:
+            # Not supported
+            return
+    else:
+        # Not supported
+        return
+
+    if not isinstance(mixing_rv.owner.op, (BernoulliRV, CategoricalRV)):
+        # Not supported
+        return
+
+    # We index the mixture components with each index element that's in the
+    # component's group
+    value_var = rv_map_feature.rv_values.pop(out_var)
```

Comment:
What if the user did not specify a value var to the mixture (because they don't want to)? This would raise a KeyError, no?

Reply:
Yes; I'll update that.